source-s3/capture.go

```diff
@@ -0,0 +1,209 @@
+
+	// Does this key hash into our shard's assigned range?
+	// We hash the bucket and the full key just to be on the safe side.
+	var hashKey = obj.bucket + "/" + fullKey
+	if !c.config.ShardRange.IncludesHwHash([]byte(hashKey)) {
+		return nil
+	}
+
+	var skipRecords uint64
+	var prevState = c.state[obj.relativeKey]
+	// Does our state indiate that we've already imported this object?
+	if prevState != nil && prevState.ETag == obj.etag {
+		if prevState.Complete {
+			return nil
+		} else {
+			skipRecords = prevState.RecordCount
+		}
+	}
+
+	var getInput = s3.GetObjectInput{
+		Bucket: &obj.bucket,
+		Key:    &fullKey,
+	}
+	var resp, err = c.client.GetObject(&getInput)
+	if err != nil {
+		return fmt.Errorf("getObject failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	var parseConfig = c.makeParseConfig(obj.relativeKey, resp.ContentType, resp.ContentEncoding)
+	configFile, err := ioutil.TempFile("", "parser-config-*.json")
+	if err != nil {
+		return fmt.Errorf("creating parser config temp file: %w", err)
+	}
+	defer os.Remove(configFile.Name())
+	parseConfig.WriteToFile(configFile)
+
+	var docCount uint64 = 0
+	err = parser.ParseStream(ctx, configFile.Name(), resp.Body, func(data json.RawMessage) error {
+		docCount++
+		// Do we need to skip this record because it's already been ingested, as indicated by the
+		// offset in the state?
+		if docCount > skipRecords {
+			return onResult(readResult{
+				bucket:      c.bucket,
+				streamID:    c.streamID,
+				relativeKey: obj.relativeKey,
+				record:      data,
+				state: &objectState{
+					ETag:         *resp.ETag,
+					LastModified: *resp.LastModified,
+					RecordCount:  docCount,
+					Complete:     false,
+				},
+			})
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return onResult(readResult{
+		bucket:      c.bucket,
+		streamID:    c.streamID,
+		relativeKey: obj.relativeKey,
+		state: &objectState{
+			ETag:         *resp.ETag,
+			LastModified: *resp.LastModified,
+			RecordCount:  docCount,
+			Complete:     true,
+		},
+	})
+}
+
+func (c *stream) makeParseConfig(relativeKey string, contentType *string, contentEncoding *string) *parser.Config {
+	var parseConfig = parser.Config{}
+	if c.config.Parser != nil {
+		parseConfig = c.config.Parser.Copy()
+	}
+
+	parseConfig.Filename = relativeKey
+	if contentType != nil {
+		parseConfig.ContentType = *contentType
+	}
+	if contentEncoding != nil {
+		parseConfig.ContentEncoding = *contentEncoding
+	}
+	// If the user supplied a location for this, then we'll use that. Otherwise, use the default
+	if parseConfig.AddRecordOffset == "" {
+		parseConfig.AddRecordOffset = defaultRecordOffsetLocation
+	}
+	if parseConfig.AddValues == nil {
+		parseConfig.AddValues = make(map[parser.JsonPointer]interface{})
+	}
+	parseConfig.AddValues[sourceFilenameLocation] = relativeKey
+	return &parseConfig
+}
+
+const sourceFilenameLocation = "/_meta/sourceFile"
+const defaultRecordOffsetLocation = "/_meta/recordOffset"
```

Comment:
"/_meta/source/record" and "/_meta/source/file" ? Consolidate source attributes into a well-known property seems a good idea.

Reply:
Yeah, agreed :+1: